06.Lesson_ThinkingProperties/ThinkingPropertiesExercise/MeetingScheduler/meetingscheduler.spec

```diff
@@ -0,0 +1,210 @@
+    numParticipants2After = getNumOfParticipents(meetingId2);
+    assert (status2Before == status2After, "status changed");
+    assert (startTime2Before == startTime2After, "startTime changed");
+    assert (endTime2Before == endTime2After, "endTime changed");
+    assert (numParticipants2Before == numParticipants2After, "numParticipants changed");
+}
+
+
+// TODO: i don't know how to compare reachability of 2 function calls (1 by the organizer, the other
+// by the non-organizer using the `storage` syntax. Gotta ask Mike)
+// all function calls that succeed for a non-organizer would all succeed for an organizer
+// rule nonOrganizerCallsAlwaysSucceedForOrganizer(address nonOrganizer, address organizer) {
+//     env e; calldataarg args;
+
+//     require e.msg.sender == organizer;
+
+
+//     // I wanna check, if we start in the same state, and then consider 2 possibilities:
+//     // 1. an organizer calls a function
+//     // 2. a non-organizer calls a function
+//     // then there should be no instances where non-organizer succeeds, and organizer fails
+//     // another way of saying this is that the set of successful function calls for organizer
+//     // is a superset of the non-organizer's set of successful function calls
+
+
+//     // take a snapshot of the state so we can compare 
+//     storage initial = lastStorage;
+// }
+
+// after a transition to the ENDED (status = 3), then block.timestamp > endTime
+rule stateOfEnded(uint256 meetingId) {
+    env e;
+
+    uint256 stateBefore = getStartTimeById(meetingId);
+    // we want to ensure we're in the right initial state, which if
+    // we're trying to transition to ENDED means we must be in the STARTED
+    // (status = 2) state
+    require stateBefore == 2;
+
+    endMeeting(e, meetingId);
+
+    uint256 endTimeAfter = getEndTimeById(meetingId);
+    uint8 stateAfter = getStateById(meetingId);
+
+    assert (stateAfter == 3, "unexpected state");
+    assert (e.block.timestamp > endTimeAfter, "block.timestamp is before the end time");
+}
+
+rule stateOfEndedParametric(uint256 meetingId) {
+    env e;
+    method f;
+    calldataarg args;
+
+    require getStateById(meetingId) != 3;
+
+    f(e, args);
+    uint8 stateAfter = getStateById(meetingId);
+
+    // ensure that the status transitioned to ENDED
+    require stateAfter == 3;
+
+    uint256 endTimeAfter = getEndTimeById(meetingId);
+
+    assert (e.block.timestamp >= endTimeAfter, "block.timestamp is before the end time");
+}
+
+// once a meeting is in the ENDED state, every non-view function called
+// on it should revert (ENDED should be a terminal state)
+rule endStateShouldCauseFailureForAllNonViewFunctions(uint256 meetingId, uint256 startTime, uint256 endTime) {
+    method f;
+
+    require f.isView == false;
+
+    uint8 stateBefore = getStateById(meetingId);
+    require stateBefore == 3;
+
+    callFunctionWithParams(f, meetingId, startTime, endTime);
+
+    assert false;
```

Comment:
You can implement this rule like this:
```
rule endStateShouldCauseFailureForAllNonViewFunctions(method f, uint256 meetingId, uint256 startTime, uint256 endTime) filtered { f -> !f.isView } {
    uint8 stateBefore = getStateById(meetingId);
    require stateBefore == 3;

    callFunctionWithParams@withrevert(f, meetingId, startTime, endTime);

    assert lastReverted;
}

Reply:
done

Reply:
Wait, you can call CVL functions `@withrevert`?  I didn't know that!